pkg/agent/kubeagent.go

```diff
@@ -126,3 +169,265 @@ func (k *KubeAgent) GetFactories(_ context.Context) (otelcol.Factories, error) {
+		c.logger.Info("restarting mw-agent")
+		if err := c.restartKubeAgent(DaemonSet); err != nil {
+			c.logger.Error("error restarting mw-agent daemonset", zap.Error(err))
+			return err
+		}
+	}
+
+	if apiResponse.Rollout.Deployment {
+		c.logger.Info("restarting mw-agent")
+		if err := c.restartKubeAgent(Deployment); err != nil {
+			c.logger.Error("error restarting mw-agent deployment", zap.Error(err))
+			return err
+		}
+	}
+
+	return err
+}
+
+// restartKubeAgent rewrites the configmaps and rollout restarts agent's data scraping components
+func (c *KubeAgentMonitor) restartKubeAgent(componentType ComponentType) error {
+
+	updateConfigMapErr := c.updateConfigMap(componentType)
+	if updateConfigMapErr != nil {
+		return updateConfigMapErr
+	}
+
+	rolloutRestartErr := c.rolloutRestart(componentType)
+	if rolloutRestartErr != nil {
+		return rolloutRestartErr
+	}
+
+	return nil
+}
+
+func (c *KubeAgentMonitor) SetClientSet() {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		c.logger.Error("Error getting in-cluster config", zap.String("error", err.Error()))
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		c.logger.Error("Error creating Kubernetes client", zap.String("error", err.Error()))
+	}
+
+	c.Clientset = clientset
+}
+
+// rolloutRestart reloads the k8s components
+// based on component type
+func (c *KubeAgentMonitor) rolloutRestart(componentType ComponentType) error {
+
+	switch componentType {
+	case DaemonSet:
+		daemonSet, err := c.Clientset.AppsV1().DaemonSets(c.AgentNamespace).Get(context.TODO(), c.Daemonset, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		daemonSet.Spec.Template.ObjectMeta.Labels[Timestamp] = fmt.Sprintf("%d", metav1.Now().Unix())
+		_, err = c.Clientset.AppsV1().DaemonSets(c.AgentNamespace).Update(context.TODO(), daemonSet, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+
+	case Deployment:
+		deployment, err := c.Clientset.AppsV1().Deployments(c.AgentNamespace).Get(context.TODO(), c.Deployment, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		deployment.Spec.Template.ObjectMeta.Labels[Timestamp] = fmt.Sprintf("%d", metav1.Now().Unix())
+		_, err = c.Clientset.AppsV1().Deployments(c.AgentNamespace).Update(context.TODO(), deployment, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// updateConfigMap gets the latest configmap from Middleware backend and updates the k8s configmap
+// based on component type
+func (c *KubeAgentMonitor) updateConfigMap(componentType ComponentType) error {
+
+	u, err := url.Parse(c.APIURLForConfigCheck)
+	if err != nil {
+		return err
+	}
+
+	baseURL := u.JoinPath(apiPathForYAML)
+	baseURL = baseURL.JoinPath(c.APIKey)
+	params := url.Values{}
+	params.Add("platform", "k8s")
+	params.Add("component_type", componentType.String())
+	params.Add("host_id", os.Getenv("MW_KUBE_CLUSTER_NAME"))
+	params.Add("cluster", os.Getenv("MW_KUBE_CLUSTER_NAME"))
```

Comment:
Get value of env variable once and store it in a variable. Use that variable everywhere

```
clusterName := os.Getenv("MW_KUBE_CLUSTER_NAME")
params.Add("host_id", clusterName)
params.Add("cluster", clusterName)
```